xonsh/environ.py

```diff
@@ -7,62 +7,188 @@
 import builtins
 import platform
 import subprocess
+from datetime import datetime
+from functools import partial
 from warnings import warn
 
 from xonsh.tools import TERM_COLORS
 
-def current_branch(cwd=None):
-    """Gets the branch for a current working directory. Returns None
-    if the cwd is not a repository.  This currently only works for git, 
-    bust should be extended in the future.
-    """
-    branch = None
-    cwd = os.getcwd() if cwd is None else cwd
+class PromptFormatter(dict):
+    '''
+    Base class that implements utilities for PromptFormatters.
 
-    # step out completely if git is not installed
-    try:
-        binary_location = subprocess.check_output(['which', 'git'], cwd=cwd,
-                                    stderr=subprocess.PIPE,
-                                    universal_newlines=True)
-        if not binary_location:
-            return branch
-    except subprocess.CalledProcessError:
-        return branch
+    You should read the docs for this class to learn how to write your own
+    Prompt Formatter.  However, you most likely want to inherit from
+    DefaultPromptFormatter if you are writing a custom formatter.
+
+    Custom Prompt Formatters will need to implement their own __init__()
+    method that adds entries to several dicts depending on the needs of the
+    variables they add.  The Formatter may also implement methods to implement
+    those variables (although very simple variables can be written with
+    pre-existing functions.)
+
+    Example:
```

Comment:
>
> sphinx has to run the code through the python compile step so we cannot
> use a version of sphinx-build running on python2. Must run it via python3.
> I noticed that your scripts/xonsh also used a /usr/bin/python shebang
> instead of /usr/bin/python3 so maybe you have your system setup so that
> python maps to python3 (perhaps just when running certain applications)?
> That would explain why you don't need that but other people do (I had to
> change the shebang line to run from a checkout as well. The shebang is
> properly rewritten when installed if python3 seutp.py install is used).
>

I probably do.  Adding the trailing 3 is probably reasonable for the
moment.


> Looks like python3 support in cloud_sptheme on pypi is broken. They have
> it fixed in source control but they haven't made a release since 2013 so
> they haven't pushed the fix out where it's easily available. I've locally
> my copy so that it runs.
>
Ahh, you should probably complain to them :)


> autodoc isn't doing a good job of organizing the docs for the new classes.
>
What do you mean by this?  In terms of the ordering? I think that in sphinx
you can switch the ordering around.

> https://toshio.fedorapeople.org/environ.html what do you think of this
> plan:
>
>    - Pull the formatters out of environ into their own file.
>
> I am not a huge fan of that, yet. I don't think that there is enough code
to justify that.

>
>    - Use autodoc of specific methods of the formatter classes instead of
>    autodoc of the whole file.
>
> kinda against this because it makes it much more annoying to maintain the
docs.

>
>    - * Turn off inherited-members for the new formatter documentation.
>
> I'd be happy with that.

>
>    - Either create a new rst page in api for prompt formatters or add it
>    as a separate section o nteh environ page (sine format_prompt() is there).
>
> I'd be happier with a separate section.

>
>
> Question: Do you still want documentation of the prompt variables in
> format_prompt? I think it may be more logical to document that in
> DefaultPromptFormatter (probably in the class docstring since not all
> format variables are implemented via a method.)
>
 Yes, I think it makes sense to move most of the docstring contents.

> Will need to point the link from the tutorial section to
> DefaultPromptFormatter as well.
>
Sure.

> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scopatz/xonsh/pull/125/files#r26908585>.
>